src/build_helper/src/git.rs

```diff
@@ -27,145 +28,195 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
+///
+/// `target_paths` should be a non-empty slice of paths (git `pathspec`s) relative to `git_dir`
+/// whose modifications would invalidate the artifact.
+/// Each pathspec can also be a negative match, i.e. `:!foo`. This matches changes outside
+/// the `foo` directory.
+/// See https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec
+/// for how git `pathspec` works.
+///
+/// The function behaves differently in CI and outside CI.
+///
+/// - Outside CI, we want to find out if `target_paths` were modified in some local commit on
+/// top of the latest upstream commit that is available in local git history.
+/// If not, we try to find the most recent upstream commit (which we assume are commits
+/// made by bors) that modified `target_paths`.
+/// We don't want to simply take the latest master commit to avoid changing the output of
+/// this function frequently after rebasing on the latest master branch even if `target_paths`
+/// were not modified upstream in the meantime. In that case we would be redownloading CI
+/// artifacts unnecessarily.
+///
+/// - In CI, we use a shallow clone of depth 2, i.e., we fetch only a single parent commit
+/// (which will be the most recent bors merge commit) and do not have access
+/// to the full git history. Luckily, we only need to distinguish between two situations:
+/// 1) The current PR made modifications to `target_paths`.
+/// In that case, a build is typically necessary.
+/// 2) The current PR did not make modifications to `target_paths`.
+/// In that case we simply take the latest upstream commit, because on CI there is no need to avoid
+/// redownloading.
+pub fn check_path_modifications(
+    git_dir: &Path,
     config: &GitConfig<'_>,
-    git_dir: Option<&Path>,
-) -> Result<String, String> {
-    let mut git = Command::new("git");
-    if let Some(git_dir) = git_dir {
-        git.current_dir(git_dir);
+    target_paths: &[&str],
+    ci_env: CiEnv,
+) -> Result<PathFreshness, String> {
+    assert!(!target_paths.is_empty());
+    for path in target_paths {
+        assert!(Path::new(path.trim_start_matches(":!")).is_relative());
     }
-    git.args(["config", "--local", "--get-regex", "remote\\..*\\.url"]);
-    let stdout = output_result(&mut git)?;
 
-    let rust_lang_remote = stdout
-        .lines()
-        .find(|remote| remote.contains(config.git_repository))
-        .ok_or_else(|| format!("{} remote not found", config.git_repository))?;
+    let upstream_sha = if matches!(ci_env, CiEnv::GitHubActions) {
+        // Here the situation is different for PR CI and try/auto CI.
+        // For PR CI, we have the following history:
+        // <merge commit made by GitHub>
+        // 1-N PR commits
+        // upstream merge commit made by bors
+        //
+        // For try/auto CI, we have the following history:
+        // <**non-upstream** merge commit made by bors>
+        // 1-N PR commits
+        // upstream merge commit made by bors
```

Comment:
Both of these seem like a bit of a simplistic view of history? In particular it's not accounting for the multi-parent case (or at least doesn't seem to reference it), which feels surprising to me. Maybe worth mentioning in the comments somehow?

In particular we say "latest" commit and we should probably define that -- is that closest to $now (by time), closest topologically, etc.

Reply:
Could you please clarify what do you mean by multi-parent case? I think that on CI, `HEAD` always has exactly two parents, the first one is the parent commit from `master`, and the second is the PR HEAD commit. Or not?

Reply:
Just two parents is enough :) The history here is being outlined as linear, whereas in practice it's at least at least two children, as you say.

I think just mentioning that would be enough, I don't think any of the logic here is actually wrong due to this. But as-is we're implying a linear history, whereas (as noted below) we actually have potentially more complex situation -- e.g., the PR commits may have parents of bors merge commits too (i.e., be merge commits themselves). I _think_ those are never more recent than the one referenced here, so our current strategy doesn't care.

Reply:
Created a small ASCII diagram to make the description more accurate.